### PR TITLE
[backend/frontend] Add user and group confidence level (#4304)

### LINKED ISSUE #4304: Add Confidence level on Users
## Use case

Users have a Max Confidence level that will be used when creating Knowledge. 
As connectors, stream and feeds used a user to upsert data, confidence level in their config is no longer needed. They will use the max confidence level of the users they are associated with.

In User's edition form, a dedicated panel allows to set the max confidence level

### No migration

We do not know what is the value used in the connectors configuration files of our users, and we cannot know which user is used for a connector, thus we cannot magically migrate the existing users with a smart max confidence value.
For instance, migrating users to 100 would be a very dangerous move as connectors would suddenly be able to overwrite all data in the platform. 

For this reason, no migration will be executed. Platform admins are expected to configure their confidence policy throughout their platform.

### Subtasks:
* Managing a “max confidence level” for a user
* Enforce specific rights to manage confidence levels of users (capability SETACCESS)
* [Product] Anticipate the breaking change using Filigran's communication tools

### Linked to issues
* #4901 
* #5692 
* #5694
* #5695
* #5697
* #5696

### Next steps
* Using max confidence level of the associated user when upserting data
* Taking into account the max confidence level of users when they try to modify the confidence level of an Object

### Further comments
Note this feature will then be leveraged for #2517 in the future.

### LINKED ISSUE #4901: Managing max confidence level at Group level
## Use case

In order to easily apply and manage max confidence level to multiple users, we will implement max confidence level to Group. Max confidence level of users will inherit ones of Group level if none has been define at user level. If the user is part of multiple groups, the more restrictive max confidence level will be take into account.

### Subtasks:
* Managing a “max confidence level” for a Group
* Implements the "user effective max confidence level", as the computed value depending on the groups and the user's own configuration of confidence.

### Linked to
* #4304
* #5692
* #5694
* #5695 
* #5697
* #5696

### LINKED ISSUE #5692: Add admin tools for managing Groups and Users max confidence levels
The introduction of confidence level to Users and Groups will leave existing platforms in a potentially unstable state: the existing users and groups are not migrated to any confidence value, and it will be mandatory to have a proper confidence policy implemented to each platform according to the CTI team needs.

There will be a delay between the introduction of confidence levels and their usage in the platform ingestion mechanisms. Admins will have time to setup their confidence policy, then update to the next version of OpenCTI that leverages it into the platform internals.

This issue i about providing platform admins with tools to facilitate greatly the implementation of their confidence policy across users and groups.

The easiest way to tackle this problem is to configure the existing groups with a max confidence level.
As Users are members of groups (if not, it means they do not have any permissions in the app), they will inherit the group(s) confidence level properly.

### subtasks
* add a popup dialog at the start of the platform (login or refresh) that warns platform admins that there is problems with their confidence policy. Specifically, we warn about groups without confidence level set. No issue, no popup.
* Highlight in the UI issues with users and groups without confidence level, with alerts, icons, and tooltips
* add the ability to select groups when creating a user (if groups are specified this way, they overrides the default groups that have been set in the platform)
* show group max confidence level in the Select UI when creating a user
* show groups confidence level in the user edition > group tab

### linked to
* #4304 
* #4901 
* #5694
* #5695 
* #5697
* #5696

### PR DESCRIPTION
**Pull request by labo-flg**
### Proposed changes
This PR addresses the following issues:
* Closes #4304
* Closes #4901
* Closes #5692  

Have a look at each issue description for details.

### Code review highlights
The key points to review are backend additions and changes, the model definition.
A lot of stuff in the frontend besides the core job (refactor in TS, solving react warnings) make the review difficult, I'm sorry.

### How can you test this PR?
Here are some scenarios to test:

* start the platform on this branch -> you shall see the critical alert dialog, inviting you to edit all your groups

* Create a user without groups -> they should be created with default groups
* Create a user with groups -> they should be created with the specified groups, and not the default group

* Create a user without a confidence level and no group 
  * it should be displayed as in error in the various view (list, details, list of users in a group details view) -> icon + tooltip
  * edit the user : the message above the confidence level field shall be explicit (no group to inherit from)
  * edit the user, tab groups : the confidence of each group shall be displayed, with red icon and tooltip if not defined
  * edit the user : set a group in the groups tab, go back to overview -> the message above the confidence level field shall be explicit (inherited value)
  * edit the user : set the user's confidence level -> -> the message above the confidence level field shall be explicit (user value)
* Create a user without a confidence level and no group -> it should be displayed as in error in the various view (list, details, list of users in a group details view)
* Create a user with a confidence level -> it should be displayed in the various view (list, details, list of users in a group details view) -> tooltip shall state it's coming from user
* Create a user without a confidence level but with a group -> it should be displayed in the various view (list, details, list of users in a group  details view) -> tooltip shall state it's coming from a specific group - link clickable in tooltip

* Create a group, you must select a confidence level -> value is then displayed correctly in list and details
* Open a group without confidence level -> explicit alert in details -> disappears when edited

* all groups are set with a confidence level -> refresh page -> no more alert dialog

If you wish to artificially nuke your group confidence level for testing purpose, a GQL query on playground is possible (there is no validation implemented yet - see #5696).

```graphql
mutation GroupEditionOverviewFieldPatchMutation(
  $id: ID!
  $input: [EditInput]!
) {
  groupEdit(id: $id) {
    fieldPatch(input: $input) {
      id
      group_confidence_level {
        max_confidence
      }
    }
  }
}
```
```json
{
  "id": "4b762ee9-b2da-49fe-99d7-dfd6b9e1fe33",
  "input": {
    "key": "group_confidence_level",
    "value": [null]
  }
}

```

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

Now that the new schema definition is in place, the job done on the subject of user's confidence level has been fully reworked.
Thus this PR invalidates the previous PR on the subject: #5283 #5292 #5323

### CODE REVIEW

File: opencti-platform/opencti-front/src/private/components/settings/users/UserEditionOverview.tsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserEditionOverview.tsx]: This is how we can now field patch the object properly.

FYI it also works with overrides with payload like:

```json
{
  "id": "26a02498-18e4-4b51-af18-44e5d8d6d5d9",
  "input": {
  "key": "user_confidence_level",
    "object_path": "/user_confidence_level/overrides/0/max_confidence",
    "value": 7
}
```

File: opencti-platform/opencti-graphql/src/migrations/1704792752287-user-confidence-level.js
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/migrations/1704792752287-user-confidence-level.js]: This mapping is the one generated by the platform based on the schema (I've checked in kibana).

File: opencti-platform/opencti-graphql/src/modules/attributes/internalObject-registrationAttributes.ts
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/modules/attributes/internalObject-registrationAttributes.ts]: I'd like a review on this definition please @richard-julien 🙏

**Comment by labo-flg** [comment]
I'm able to send through the API a partial update like

```json
{
  "id": "<id>",
  "input": {
  "key": "user_confidence_level",
    "value": {
    "max_confidence": 96
  }
}
```

Which will update the `user_confidence_level` with overrides to null, causing GQL error at query time as this field is supposedly never null.
I'll check the model, maybe around bad `mandatoryType`

File: opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx]:
```suggestion
type UserConfidenceLevelProps = {
```

File: opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx]:
```suggestion
const UserConfidenceLevel: React.FC<UserConfidenceLevelProps> = ({ userConfidenceLevel }) => {
```

File: opencti-platform/opencti-graphql/config/default.json
Comment (lndrtrbn) [inline opencti-platform/opencti-graphql/config/default.json]: Do we keep 100 here as the migration set 0 by default ?

**Review by lndrtrbn — COMMENTED** [review]
Seems good to me, but I'll let someone else have a look at migration and attributes registrations 😅

File: opencti-platform/opencti-graphql/src/modules/attributes/internalObject-registrationAttributes.ts
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/modules/attributes/internalObject-registrationAttributes.ts]: reviewed in pair. I'd like to make sure the update cannot be done on incomplete objects, like 

```json
{
  "id": "26a02498-18e4-4b51-af18-44e5d8d6d5d9",
  "input": {
  "key": "user_confidence_level",
    "value": {
      "max_confidence": 96
  }
}
```
This is currently accepted and removes the overrides, which throws a GQL non-nullable field error on query

File: opencti-platform/opencti-graphql/config/default.json
Comment (labo-flg) [reply opencti-platform/opencti-graphql/config/default.json]: Yes we keep it there. When adding a new user, we still want it to be 100 (this user is new and supposedly not linked to a connector). 

Later, the confidence level will be set on the group level and we will effectively use the lowest confidence level among the user's groups and own level.

**Comment by labo-flg** [comment]
> Seems good to me, but I'll let someone else have a look at migration and attributes registrations 😅

Thanks !

**Comment by labo-flg** [comment]
back to draft : I need to make sure partial edit payload are not accepted

**Comment by labo-flg** [comment]
We need to add schema validation when patching a field.
This is significant work that deserve a dedicated branch/PR.

Everything in this PR is ok, but we need this validation before merging to master.
I will use this branch as a feature branch, and merge into it the validation.

EDIT: validation will be done in another issue (#5696) , this branch will still serve as feature branch

File: opencti-platform/opencti-front/src/components/InputSliderField.tsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/components/InputSliderField.tsx:61]: this align the 2 inputs (input number + scale Selector) to the top, keeping alignments when the input number shows an error label below.

File: opencti-platform/opencti-front/src/private/Index.tsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/Index.tsx:81]: New popup dialog for admins when the backend has some "critical alerts" to show.
This is for warning about groups with null confidence levels, but could be extended to other types of alerts

File: opencti-platform/opencti-front/src/private/Root.tsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/Root.tsx:39]: alert details shall depends on alert type... current implem is a bit rough on the edges but we had to move quickly about this.

File: opencti-platform/opencti-front/src/private/components/common/form/GroupField.tsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/common/form/GroupField.tsx]: old GroupField Select, refactored to typescript, and added opt prop showConfidence to add the group confidence to the label after the name of the group.

File: opencti-platform/opencti-front/src/private/components/settings/Users.jsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/settings/Users.jsx:62]: effective confidence level is a computed value backend side, not an attribute in DB (depends on user's confidence + his groups confidence)

we cannot craft a query to do this computation, so we cannot paginate and sort properly this value.

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionOverview.tsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionOverview.tsx:137]: the else statement could work in both cases (existing confidence or not) right now, but we will implement soon the overrides definition and edition (6.1). By then we'll need to patch with `object_path` anyway.

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx:63]: a technical field (not to submit) associated with a `when`. Very useful for validation conditioned on other values!

File: opencti-platform/opencti-graphql/src/domain/user.js
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/domain/user.js:1463]: function split for easy unit tests

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql
Comment (labo-flg) [inline opencti-platform/opencti-graphql/config/schema/opencti.graphql:1074]: this is where the critical alert system is a bit limited, not well generalized. details shall depends on type.

I did not have time to overengineer this system, honestly it's a quick bandaid to help admins prepare for 6.0.

Comment (richard-julien) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql:1074]: Not a problem, thats purely dynamic and so easy to change.

File: opencti-platform/opencti-front/lang/en.json
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/en.json]:
```suggestion
  "alert_GROUP_WITH_NULL_EFFECTIVE_LEVEL": "Starting with OpenCTI 6.0, a max confidence level will be added to groups and users. It will provide platform administrators with more control on the impact of imported data and prevent any unwanted update on data. Please read {link_blogpost} for more information.\n \nOn this platform, \"max confidence level\" of some groups are currently undefined. Their members will not be able to create any data in version 6.X of the platform. In order to prevent that, make sure that all groups have a max confidence level set. More info in {link_blogpost}. Be particularly mindful of users associated to Connectors, Feeds and Streams; you can set individual confidence level to override group confidence level if needed. Join {link_slack} if you need advices.",
```
Feel free to refuse, "if need be" sound weird to me, but English is not my native language.

File: opencti-platform/opencti-front/lang/fr.json
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/fr.json:2521]:
```suggestion
  "This value is overridden for the following entity types": "Cette valeur est remplacée pour les types d'entités suivantes",
```
I think that "entité" is feminine.

File: opencti-platform/opencti-front/lang/fr.json
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/fr.json:2539]: question: I don't see this in the "en" file, is it normal ?

File: opencti-platform/opencti-front/lang/en.json
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/lang/en.json]: why has this translation been removed ?

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupLine.tsx
Comment (aHenryJard) [inline opencti-platform/opencti-front/src/private/components/settings/groups/GroupLine.tsx:66]: question: This is the first time I see this empty <> in code, is it intended ?

**Review by aHenryJard — COMMENTED** [review]
I just did a "quick read" review so I share it, I will take take time for a deeper one later.

File: opencti-platform/opencti-front/lang/en.json
Comment (labo-flg) [reply opencti-platform/opencti-front/lang/en.json]: In must have executed an extract script in my branch, pre-rebase.
I'll check and restore the missing keys.

File: opencti-platform/opencti-front/lang/fr.json
Comment (labo-flg) [reply opencti-platform/opencti-front/lang/fr.json:2521]: indeed ! FYI it's Deepl auto translate

Comment (labo-flg) [reply opencti-platform/opencti-front/lang/fr.json:2521]: hm actually it's "types suivants"

File: opencti-platform/opencti-front/lang/fr.json
Comment (labo-flg) [reply opencti-platform/opencti-front/lang/fr.json:2539]: it's in the file. maybe not at the same line, as files other than en are generated.

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupLine.tsx
Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/settings/groups/GroupLine.tsx:66]: it's a react fragment, a virtual organization of component (it's interpreted as a single react node containing all the children, but at render time there is no actual tag (div or anything else).
`<>` is a shortcut to <React.Fragment>

Comment (lndrtrbn) [reply opencti-platform/opencti-front/src/private/components/settings/groups/GroupLine.tsx:66]: And it is used because here you can have only one child at root. So as we want multiple we use an empty React fragment as root and imbricate our multiple elements inside. The plus to use this instead of using a div is that by the end we don't add anything in the generated DOM

File: opencti-platform/opencti-front/lang/en.json
Comment (labo-flg) [reply opencti-platform/opencti-front/lang/en.json]: I've done a review, all keys not related to confidence level have been restored (there was a couple)

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupConfidenceLevel.tsx
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/groups/GroupConfidenceLevel.tsx:27]: Is it a normal TODO or a missed one ?

File: opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx:30]: TODO later ?

File: opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx:35]: FIXME later ?

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupConfidenceLevel.tsx
Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/settings/groups/GroupConfidenceLevel.tsx:27]: yes normal, it will come in 6.1 with #5694

File: opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx
Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx:30]: yes normal, it will come in 6.1 with #5694

File: opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx
Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx:35]: I did not have time to investigate, but it's probably related to how we poorly handle the graphql cache records

Warning: RelayModernRecord: Invalid record update, expected both versions of record `88ec0c6a-13ce-5e39-b486-354fe4a7084f` to have the same `__typename` but got conflicting types `MeUser` and `User`. The GraphQL server likely violated the globally unique id requirement by returning the same id for different objects.

Another whole issue to address, certainly not in this PR.

Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserConfidenceLevel.tsx:35]: I'm editing myself, which collides with `MeUser` query for instance. The record gets "updated" wrongly somehow.

**Review by lndrtrbn — APPROVED** [review]

**Pull request merged**
